---
title: "Class 18: Pertussis Mini Project"
author: "Aileen Andrade (PID A17033749)"
date: "03-09-2025"
format: pdf
---

Pertussis (a.k.a.) Whooping Cough is a deadly lung infection caused by the the bacteria B. Pertussis. 

The CDC tracks Pertussis cases around the U.S.
https://tinyurl.com/pertussiscdc

We can "scrape" this data using the R **datapasta** package.

```{r, include=FALSE}
cdc <- data.frame(
  Year = c(1922L,1923L,1924L,1925L,
           1926L,1927L,1928L,1929L,1930L,1931L,
           1932L,1933L,1934L,1935L,1936L,
           1937L,1938L,1939L,1940L,1941L,1942L,
           1943L,1944L,1945L,1946L,1947L,
           1948L,1949L,1950L,1951L,1952L,
           1953L,1954L,1955L,1956L,1957L,1958L,
           1959L,1960L,1961L,1962L,1963L,
           1964L,1965L,1966L,1967L,1968L,1969L,
           1970L,1971L,1972L,1973L,1974L,
           1975L,1976L,1977L,1978L,1979L,1980L,
           1981L,1982L,1983L,1984L,1985L,
           1986L,1987L,1988L,1989L,1990L,
           1991L,1992L,1993L,1994L,1995L,1996L,
           1997L,1998L,1999L,2000L,2001L,
           2002L,2003L,2004L,2005L,2006L,2007L,
           2008L,2009L,2010L,2011L,2012L,
           2013L,2014L,2015L,2016L,2017L,2018L,
           2019L,2020L,2021L, 2022L, 2024L),
  Cases = c(107473,164191,165418,152003,
                                   202210,181411,161799,197371,
                                   166914,172559,215343,179135,265269,
                                   180518,147237,214652,227319,103188,
                                   183866,222202,191383,191890,109873,
                                   133792,109860,156517,74715,69479,
                                   120718,68687,45030,37129,60886,
                                   62786,31732,28295,32148,40005,
                                   14809,11468,17749,17135,13005,6799,
                                   7717,9718,4810,3285,4249,3036,
                                   3287,1759,2402,1738,1010,2177,2063,
                                   1623,1730,1248,1895,2463,2276,
                                   3589,4195,2823,3450,4157,4570,
                                   2719,4083,6586,4617,5137,7796,6564,
                                   7405,7298,7867,7580,9771,11647,
                                   25827,25616,15632,10454,13278,
                                   16858,27550,18719,48277,28639,32971,
                                   20762,17972,18975,15609,18617,
                                   6124,2116,3044, 35493)
)
```

```{r}
head(cdc)
```

> Q1. With the help of the R “addin” package datapasta assign the CDC pertussis case number data to a data frame called cdc and use ggplot to make a plot of cases numbers over time.


```{r}
library(ggplot2)

ggplot(cdc) +
  aes(Year, Cases) +
  geom_line() 

```

> Q2. Using the ggplot geom_vline() function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?

```{r}
library(ggplot2)

ggplot(cdc) +
  aes(Year, Cases) +
  geom_line() +
  geom_vline(xintercept = 1946, col="Blue") +
  geom_vline(xintercept = 1996, col="Red") +
  geom_vline(xintercept = 2020, col="Gray") +
  geom_vline(xintercept = 2003, col="Dark green")
```

# Observations: There were high case numbers before the first wP (whole-cell) vaccine roll out in 1946 then a rapid decline in case numbers until 2004 when we have our first large-scale outbreaks of pertussis again. There is also a notable COVID related dip and recent rapid rise. 

> Q3.Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?

# After the introduction of the aP vaccine, in 1996 pertussis cases remained low initially until there was a resurgence starting around the 2000s. In 2012, there were 48,277 reported cases which was the highest since 1955. A possible explanation for the observed trend involves more sensitive PCR-based testing since the improved accuracy and sensitivity of the tests could have led to more detection as well as a decline in vaccines due to concerns. Additionally, bacterial evolution could have been a factor as well as the fact that the aP vaccine doesn't provide long-lasting immunity. 

Q. What is different about the immune response to infection if you had an older wP vaccine vs the newer aP vaccine?

## Computational Models of Immunity - Pertussis Boost CMI-PB

The CMI-PB project aims to address this key question: what is different between aP and wP individuals.

We can get all the data from this ongoing project via JSON API calls.
For this we will use the **jsonlite** package. We can install with: `install.packages("jsonlite")`
```{r}
library(jsonlite)

subject <- read_json("https://www.cmi-pb.org/api/v5_1/subject", simplifyVector = T)
```

> Q. How many individuals "subjects" are in this data set?

```{r}
nrow(subject)
```

> Q4. How many wP and aP primed individuals are in this dataset?

```{r}
table(subject$infancy_vac)
```


> Q5. How many Male and Female subjects/patients are in the dataset?

```{r}
table(subject$biological_sex)
```

> Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?

```{r}
table(subject$race, subject$biological_sex)
```
> Q7. Using this approach determine (i) the average age of wP individuals, (ii) the average age of aP individuals; and (iii) are they significantly different?


```{r}
library(lubridate)
library(dplyr)

subject$age <- today() - ymd(subject$year_of_birth) 

subject$age_years <- time_length(subject$age, "years")

ap <- subject %>% filter(infancy_vac == "aP")
round(summary(time_length(ap$age, "years")))

wp <- subject %>% filter(infancy_vac == "wP")
round(summary(time_length(wp$age, "years")))

# Check if the difference is statistically significant using a t-test
t.test(ap$age_years, wp$age_years)

```
# The average age of wP individuals is 35.83 years. The average age of aP individuals is 27.08 years. The difference is statistically significant (p-value < 2.2e-16), meaning the wP group is significantly older than the aP group.


> Q8. Determine the age of all individuals at time of boost?

```{r}
int <- ymd(subject$date_of_boost) - ymd(subject$year_of_birth)
age_at_boost <- time_length(int, "year")
head(age_at_boost)
```


> Q9 (1). With the help of a faceted boxplot or histogram (see below), do you think these two groups are significantly different?

```{r}
ggplot(subject) +
  aes(time_length(age, "year"),
      fill=as.factor(infancy_vac)) +
  geom_histogram(show.legend=FALSE) +
  facet_wrap(vars(infancy_vac), nrow=2) +
  xlab("Age in years")
```
With the help of a faceted histogram yes, the two groups are significantly different in age distrbution.

Obtain more data from CMI-PB:

```{r}
specimen <- read_json("https://cmi-pb.org/api/v5_1/specimen",
                      simplifyVector = T)

ab_data <- read_json("https://www.cmi-pb.org/api/v5_1/plasma_ab_titer",
                      simplifyVector = T)
```

```{r}
head(ab_data)
```

> Q9 (2). Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details:

I now have 3 tables of data from CMI-PB: `subject`, `specimen`, and `ab_data`. I need to "join" these tables so I will have all the info I need to work with.

For this we will use the `inner_join()` function from **dplyr** package.

```{r}
library(dplyr)

meta <- inner_join(subject, specimen)
head(meta)
```
```{r}
dim(subject)
dim(specimen)
```


```{r}
dim(meta)
```

> Q10. Now using the same procedure join meta with titer data so we can further analyze this data in terms of time of visit aP/wP, male/female etc.

Now we can join our `ab_data` table to `meta` so we have all the info we need about antibody levels.

```{r}
abdata <- inner_join(meta, ab_data)
head(abdata)
```

> Q. How many different antibody isotypes are there in this dataset?

```{r}
length(abdata$isotype)
```

> Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?

```{r}
table(abdata$isotype)
```

> Q12. What are the different $dataset values in abdata and what do you notice about the number of rows for the most “recent” dataset?

```{r}
table(abdata$antigen)
```
> Q13. Complete the following code to make a summary boxplot of Ab titer levels (MFI) for all antigens:

I want a plot of antigen levels across the whole dataset.

```{r}
library(ggplot2)

ggplot(abdata) +
  aes(MFI, antigen) +
  geom_boxplot()
```

```{r}
ggplot(abdata) +
  aes(MFI_normalised, antigen) +
  geom_boxplot()
```

# Question 13 Summary Boxplot
```{r}
igg <- abdata %>% filter(isotype == "IgG")
```

```{r}
igg_filtered <- igg %>% filter(MFI_normalised <= 75)

ggplot(igg_filtered) +
  aes(x = MFI_normalised, y = antigen) +
  geom_boxplot() + 
  facet_wrap(vars(visit), nrow = 2)
```


> Q14. What antigens show differences in the level of IgG antibody titers recognizing them over time? Why these and not others?

Antigens like FIM2/3, PT, FELD1 have quite a large range of values. Others like Measles don't show much activity. FIM2/3 and PT show differences since they're major components of the pertussis vaccines so individuals exposed to those vaccines can develop immune responses to them. Additionally, FELD1 is an antigen that has a strong immune reaction in certain individuals possibly due to previous infections or the environment. PT tends to also generate strong antibody responses in individuals with different vaccine types; measles doesn't show much activity since its antibody titers remain relatively stable as measles can induce long-lasting immunity from infection or vaccination.

> Q. Are there differences at this whole-dataset level between aP and wP?

```{r}
ggplot(abdata) +
  aes(MFI_normalised, antigen, col=infancy_vac) +
  geom_boxplot()
```

```{r}
ggplot(abdata) +
  aes(MFI_normalised, antigen, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(~infancy_vac)
```

## Examine IgG Ab titer levels

For this I need to select out just isotype IgG.

```{r}
igg <- abdata |>
  filter(isotype == "IgG")
head(igg)
```

A overview boxplot: 
```{r}
ggplot(igg) +
  aes(MFI_normalised, antigen, col=infancy_vac) +
  geom_boxplot()
```

```{r}
ggplot(igg) +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(visit), nrow=2) +
  xlim(0,75) +
  theme_bw()
```

# Another version of this plot adding infancy_vac to the faceting:


```{r}
igg %>% filter(visit != 8) %>%
ggplot() +
  aes(MFI_normalised, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  xlim(0,75) +
  facet_wrap(vars(infancy_vac, visit), nrow=2)
```
> Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“OVA”, that is not in our vaccines) and a clear antigen of interest (“PT”, Pertussis Toxin, one of the key virulence factors produced by the bacterium B. pertussis).

```{r}
library(dplyr)
library(ggplot2)

# Boxplot for OVA (Control Antigen - not in vaccine)
filter(igg, antigen == "OVA") %>%
  ggplot() +
  aes(x = MFI_normalised, col = infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw() +
  labs(title = "OVA Antigen IgG Levels", x = "MFI Normalised", y = "Density")

# Boxplot for PT (Pertussis Toxin - key virulence factor)
filter(igg, antigen == "PT") %>%
  ggplot() +
  aes(x = MFI_normalised, col = infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw() +
  labs(title = "Pertussis Toxin (PT) IgG Levels", x = "MFI Normalised", y = "Density")

# Boxplot for FIM2/3 (Fimbriae 2/3 - Pertussis Vaccine Component)
filter(igg, antigen == "FIM2/3") %>%
  ggplot() +
  aes(x = MFI_normalised, col = infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw() +
  labs(title = "Fimbriae 2/3 (FIM2/3) IgG Levels", x = "MFI Normalised", y = "Density")
```

> Q16. What do you notice about these two antigens time courses and the PT data in particular?

I notice that OVA titers remain low and stable which confirms its a control antigen; it is also not present in the pertussis vaccine so theres no immune response to it. PT titers rise, peak, and decline showing an immune response to a vaccine antigen. Lastly, the wP and aP subjects pattern is similar, both vaccine types induce a response to PT but the differences can be analyzed further. 

> Q17. Do you see any clear difference in aP vs. wP responses?

There are clear differences, aP individuals have a higher initial IgG titer response but the levels decline faster. wP individuals have more sustained IgG levels with potentially longer-lasting immune responses than aP individuals.

> Q18. Does this trend look similar for the 2020 dataset?

Digging in further to look at the time course of IgG isotype PT antigen levels across aP and wP individuals:

```{r}
## Filter to include 2021 data only
abdata.21 <- abdata |> 
  filter(dataset == "2021_dataset")

## Filter to look at IgG PT data only
pt.igg <- abdata.21 |>
  filter(isotype == "IgG",  antigen == "PT") 
  
# Plot and color by infancy_vac (wP vs aP)
  ggplot(pt.igg) +
    aes(x=planned_day_relative_to_boost,
        y=MFI_normalised,
        col=infancy_vac,
        group=subject_id) +
    geom_point() +
    geom_line() +
    geom_vline(xintercept=0, linetype="dashed") +
    geom_vline(xintercept=14, linetype="dashed") +
  labs(title="2021 dataset IgG PT",
       subtitle = "Dashed lines indicate day 0 (pre-boost) and 14 (apparent peak levels)")
```

```{r}
# Filter for the 2020 dataset
abdata.20 <- abdata %>% filter(dataset == "2020_dataset")

# Plot time-course for IgG PT antigen levels in 2020 dataset
abdata.20 %>% 
  filter(isotype == "IgG", antigen == "PT") %>%
  ggplot() +
    aes(x=planned_day_relative_to_boost,
        y=MFI_normalised,
        col=infancy_vac,
        group=subject_id) +
    geom_point() +
    geom_line() +
    geom_vline(xintercept=0, linetype="dashed") +
    geom_vline(xintercept=14, linetype="dashed") +
  labs(title="2020 dataset IgG PT",
       subtitle = "Dashed lines indicate day 0 (pre-boost) and 14 (apparent peak levels)")

```

# Yes the overall trend is similar between the 2020 and 2021 datasets. Both show peaks at day 14 post-boost followed by a decline. What is most notable is that the 2020 dataset extends for a longer duration which displays antibody levels continuing to decrease over time. Across both years, the higher response in wP individuals compared to aP individuals is also consistent.

> Q19. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE) 
```

```{r}
#meta <- inner_join(specimen, subject)
ssrna <- inner_join(rna, meta)
```

```{r}
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```

> Q20.: What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?

From the plot, IGHG1 expression (TPM values) peak around visit 4 and visit 8. This can indicate two waves of B cell activation or expansion; the first peak around visit 4 can represent the initial activation of B cells and early surge in IgG1 transcription whereas the second peak around visit 8 can be secondary activation, memory B cell responses. 

> Q21. Does this pattern in time match the trend of antibody titer data? If not, why not?

No, the pattern of IGHG1 expression over time does not exactly match the trend of antibody titer data from question 15. This is because IGHG1 gene expression peaks early around visit 4 and visit 8 and it declines quickly after the peaks. Antibody titer levels in question 15 show IgG levels rise post-boost peaking around visit 4-6. They also remain high for a long period evven when IGHG1 transcription decreases. There might be a mismatch because IGHG1 expression is needed early and B cells need to transcribe the IGHG1 gene before it can start producing IgG1 antibodies. The high levels of IgG antibody titers in question 15 represent plasma cells as they continue producing antibodies for a long time, even after IGHG1 mRNA levels drop. So while IgG antibody levels remain elevated for a longer time, IGHG1 expression peaks early and decline since plasma cells continue producing IgG antibodies without needing high IGHG1 expression. Even after IGHG1 expression decreases, titer levels stay high since antibodies persist in the bloodstream for weeks to months.

